src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/SmallComparableValueTypeFrozenSet.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+
+namespace System.Collections.Frozen
+{
+    /// <summary>Provides a frozen set to use when the item is a comparable value type, the default comparer is used, and the item count is small.</summary>
+    /// <remarks>
+    /// No hashing involved, just a linear scan through the items.  This implementation is close in nature to that of <see cref="ValueTypeDefaultComparerFrozenSet{T}"/>,
+    /// except that this implementation sorts the values in order to a) extract a max that it can compare against at the beginning of each match in order to
+    /// immediately rule out values too large to be contained, and b) early-exits from the linear scan when a comparison determines the value is too
+    /// small to be contained.
+    /// </remarks>
+    internal sealed class SmallComparableValueTypeFrozenSet<T> : FrozenSetInternalBase<T, SmallComparableValueTypeFrozenSet<T>.GSW>
+    {
+        private readonly T[] _items;
+        private readonly T _max;
+
+        internal SmallComparableValueTypeFrozenSet(HashSet<T> source) : base(EqualityComparer<T>.Default)
+        {
+            // T is logically constrained to `where T : struct, IEquatable<T>, IComparable<T>`, but we can't actually write that
+            // constraint currently and still have this be used from the calling context that has an unconstrained T.
+            // So, we assert it here instead. The implementation relies on {Equality}Comparer<T>.Default to sort things out.
+            Debug.Assert(default(T) is IEquatable<T>);
+            Debug.Assert(default(T) is IComparable<T>);
+            Debug.Assert(default(T) is not null);
+            Debug.Assert(typeof(T).IsValueType);
+
+            Debug.Assert(source.Count != 0);
+            Debug.Assert(ReferenceEquals(source.Comparer, EqualityComparer<T>.Default));
+
+            _items = source.ToArray();
+            Array.Sort(_items);
```

Comment:
Why do we need to sort the array when `FindItemIndex` doesn't use binary search to locate the item?

Reply:
FindItemIndex doesn't search at all of the item is bigger than the largest, and it stops the linear search once it finds that the item is smaller than the current item in the list. The latter requires sorting and the former benefits from it.

Reply:
The size of the collection here is limited to at most 10 items.  Binary search in general has more overhead, and it's also always going to perform the full search, whereas this will exit out the moment it finds the item can't possibly be in the collection.  We can measure, but I'd be a little surprised if it's better.

Reply:
Sounds reasonable. 10 is probably too small a cutoff to make difference.